From 5fa2363435d2c2b2efe7e79ca336e49bc55f80d6 Mon Sep 17 00:00:00 2001
From: Dmitry Zakharchenko <dmitz@disroot.org>
Date: Tue, 3 Jan 2023 23:47:14 +0200
Subject: [PATCH] cyclelayouts: sync with 0.4

---
 config.def.h |  3 +++
 dwl.c        | 19 +++++++++++++++++++
 2 files changed, 22 insertions(+)

diff --git a/config.def.h b/config.def.h
index a4f7c13d0..b3e3f993d 100644
--- a/config.def.h
+++ b/config.def.h
@@ -25,6 +25,7 @@ static const Layout layouts[] = {
 	{ "[]=",      tile },
 	{ "><>",      NULL },    /* no layout function means floating behavior */
 	{ "[M]",      monocle },
+	{ NULL,       NULL },
 };
 
 /* monitors */
diff --git a/dwl.c b/dwl.c
index 19bb6ce34..eebcc85ea 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1055,6 +1056,24 @@ cursorframe(struct wl_listener *listener, void *data)
 	wlr_seat_pointer_notify_frame(seat);
 }
 
+void
+cyclelayout(const Arg *arg)
+{
+	Layout *l;
+	for (l = (Layout *)layouts; l != selmon->lt[selmon->sellt]; l++);
+	if (arg->i > 0) {
+		if (l->symbol && (l + 1)->symbol)
+			setlayout(&((Arg) { .v = (l + 1) }));
+		else
+			setlayout(&((Arg) { .v = layouts }));
+	} else {
+		if (l != layouts && (l - 1)->symbol)
+			setlayout(&((Arg) { .v = (l - 1) }));
+		else
+			setlayout(&((Arg) { .v = &layouts[LENGTH(layouts) - 2] }));
+	}
+}
+
 void
 destroydragicon(struct wl_listener *listener, void *data)
 {
